src/pci/src/configuration.rs

```diff
@@ -1451,4 +1457,134 @@ mod tests {
+            0x12,
+            None,
+            None,
+        );
+
+        // Trying to reprogram with something less than 4 bytes (length of the address) should fail
+        assert!(pci_config
+            .detect_bar_reprogramming(BAR0_REG, &[0x13])
+            .is_none());
+        assert!(pci_config
+            .detect_bar_reprogramming(BAR0_REG, &[0x13, 0x12])
+            .is_none());
+        assert!(pci_config
+            .detect_bar_reprogramming(BAR0_REG, &[0x13, 0x12])
+            .is_none());
+        assert!(pci_config
+            .detect_bar_reprogramming(BAR0_REG, &[0x13, 0x12, 0x16])
+            .is_none());
+
+        // Writing all 1s is a special case where we're actually asking for the size of the BAR
+        assert!(pci_config
+            .detect_bar_reprogramming(BAR0_REG, &u32::to_le_bytes(0xffff_ffff))
+            .is_none());
+
+        // Trying to reprogram a BAR that hasn't be initialized does nothing
+        for reg_idx in BAR0_REG..BAR0_REG + NUM_BAR_REGS {
+            assert!(pci_config
+                .detect_bar_reprogramming(reg_idx, &u32::to_le_bytes(0x1312_4243))
+                .is_none());
+        }
+
+        // Reprogramming of a 32bit BAR
+        pci_config
+            .add_pci_bar(&PciBarConfiguration {
+                addr: 0x1000,
+                size: 0x1000,
+                idx: 0,
+                region_type: PciBarRegionType::Memory32BitRegion,
+                prefetchable: PciBarPrefetchable::Prefetchable,
+            })
+            .unwrap();
+
+        assert_eq!(
+            pci_config.detect_bar_reprogramming(BAR0_REG, &u32::to_le_bytes(0x2000)),
+            Some(BarReprogrammingParams {
+                old_base: 0x1000,
+                new_base: 0x2000,
+                len: 0x1000,
+                region_type: PciBarRegionType::Memory32BitRegion
+            })
+        );
+
+        pci_config.write_config_register(BAR0_REG, 0, &u32::to_le_bytes(0x2000));
+        assert_eq!(pci_config.read_reg(BAR0_REG) & 0xffff_fff0, 0x2000);
+
+        // Attempting to reprogram the BAR with the same address should not have any effect
+        assert!(pci_config
+            .detect_bar_reprogramming(BAR0_REG, &u32::to_le_bytes(0x2000))
+            .is_none());
+
+        // Reprogramming of a 64bit BAR
+        pci_config
+            .add_pci_bar(&PciBarConfiguration {
+                addr: 0x13_1200_0000,
+                size: 0x8000,
+                idx: 1,
+                region_type: PciBarRegionType::Memory64BitRegion,
+                prefetchable: PciBarPrefetchable::Prefetchable,
+            })
+            .unwrap();
+
+        assert_eq!(pci_config.read_reg(BAR0_REG + 1) & 0xffff_fff0, 0x1200_0000);
+        assert_eq!(
+            pci_config.bars[1].r#type,
+            Some(PciBarRegionType::Memory64BitRegion)
+        );
+        assert_eq!(pci_config.read_reg(BAR0_REG + 2), 0x13);
+        assert!(pci_config.bars[2].r#type.is_none());
```

Comment:
I would assume that these asserts are not necessary for this test. The `add_pci_bar` supposed to work properly for this test.

Reply:
That test actually tests negative cases. It ensures that we correctly handle error cases. I don't think it really matters, so I'll leave it as is.